pkg/fleets/controller.go

```diff
@@ -460,19 +461,164 @@ func (c *Controller) rollingUpdateActive(fleet *agonesv1.Fleet, active *agonesv1
 	return replicas, nil
 }
 
+func (c *Controller) cleanupUnhealthyReplicas(rest []*agonesv1.GameServerSet, fleet *agonesv1.Fleet, maxCleanupCount int32) ([]*agonesv1.GameServerSet, int32, error) {
+
+	// Safely scale down all old GameServerSets with unhealthy replicas.
+	totalScaledDown := int32(0)
+	for i, gsSet := range rest {
+		if totalScaledDown >= maxCleanupCount {
+			break
+		}
+		if gsSet.Spec.Replicas == 0 {
+			// cannot scale down this replica set.
+			continue
+		}
+		if gsSet.Spec.Replicas == gsSet.Status.ReadyReplicas {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		scaledDownCount := int32(integer.IntMin(int(maxCleanupCount-totalScaledDown), int(gsSet.Spec.Replicas-gsSet.Status.ReadyReplicas)))
+		newReplicasCount := gsSet.Spec.Replicas - scaledDownCount
+		if newReplicasCount > gsSet.Spec.Replicas {
+			return nil, 0, fmt.Errorf("when cleaning up unhealthy replicas, got invalid request to scale down %s/%s %d -> %d", gsSet.Namespace, gsSet.Name, gsSet.Spec.Replicas, newReplicasCount)
+		}
+
+		gsSetCopy := gsSet.DeepCopy()
+		gsSetCopy.Spec.Replicas = newReplicasCount
+		totalScaledDown += scaledDownCount
+		if _, err := c.gameServerSetGetter.GameServerSets(gsSetCopy.ObjectMeta.Namespace).Update(gsSetCopy); err != nil {
+			return nil, totalScaledDown, errors.Wrapf(err, "error updating gameserverset %s", gsSetCopy.ObjectMeta.Name)
+		}
+		c.recorder.Eventf(fleet, corev1.EventTypeNormal, "ScalingGameServerSet",
+			"Scaling inactive GameServerSet %s from %d to %d", gsSetCopy.ObjectMeta.Name, gsSet.Spec.Replicas, gsSetCopy.Spec.Replicas)
+
+		rest[i] = gsSetCopy
+	}
+	return rest, totalScaledDown, nil
+}
+
+func (c *Controller) rollingUpdateRestFixedOnReady(fleet *agonesv1.Fleet, active *agonesv1.GameServerSet, rest []*agonesv1.GameServerSet) error {
+	if len(rest) == 0 {
+		return nil
+	}
+
+	// Look at Kubernetes Deployment util ResolveFenceposts() function
+	r, err := intstr.GetValueFromIntOrPercent(fleet.Spec.Strategy.RollingUpdate.MaxUnavailable, int(fleet.Spec.Replicas), false)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing MaxUnavailable value: %s", fleet.ObjectMeta.Name)
+	}
+	if r == 0 {
+		r = 1
+	}
+	if r > int(fleet.Spec.Replicas) {
+		r = int(fleet.Spec.Replicas)
+	}
+	unavailable := int32(r)
+
+	totalAlreadyScaledDown := int32(0)
+
+	totalScaleDownCount := int32(0)
+	// Check if we can scale down.
+	allGSS := append(rest, active)
+	readyReplicasCount := agonesv1.GetReadyReplicaCountForGameServerSets(allGSS)
+	minAvailable := fleet.Spec.Replicas - unavailable
+
+	// Check if we are ready to scale down
+	allPodsCount := agonesv1.SumSpecReplicas(allGSS)
+	newGSSUnavailablePodCount := active.Spec.Replicas - active.Status.ReadyReplicas
+	maxScaledDown := allPodsCount - minAvailable - newGSSUnavailablePodCount
+
+	if maxScaledDown <= 0 {
+		return nil
+
+	}
+	rest, _, err = c.cleanupUnhealthyReplicas(rest, fleet, maxScaledDown)
+	if err != nil {
```

Comment:
Could you please leave a comment explaining why do you just log an error here and do not return it?